airbyte_cdk/sources/streams/http/expiring_dictionary.py

```diff
@@ -0,0 +1,48 @@
+import time
+
+
+class ExpiringDictionary:
+    def __init__(self, expiration_time: int=2):
+        self._store = {}
+        self._expiration_time = expiration_time
+
+    def remove_expired_keys(self):
+        """Remove expired keys from the dictionary."""
+        current_time = time.time()
+        expired_keys = []
+        for key, (value, timestamp) in self._store.items():
+            if current_time - timestamp >= self._expiration_time:
+                expired_keys.append(key)
+            else:
+                # After Python 3.7, items will be presented in the order they were added.
```

Comment:
Based on this, I would assume that it is expiring the entries based on the "created_at" and not the "updated_at" which seems to be an issue as one request could retry for a long time while the others might be successful and we don't need to retry them.

Two proposals:
* Keep the logic based on the time find a way to order this by updated time without too much computation power. If you'd ask me right now, I would think that this is not an easy problem to solve but I'm not opposed in trying to solve it
* In `HttpClient._handle_error_resolution`, if error_resolution.response_action not in self._ACTIONS_TO_RETRY_ON, then evict the entry from the cache where `self._ACTIONS_TO_RETRY_ON = {ResponseAction.RETRY, ResponseAction.RATE_LIMITED}`

Reply:
I think [this commit](https://github.com/airbytehq/airbyte-python-cdk/pull/92/commits/5a4a6235e2513d8c508f65a5dfad8e715dde5db0) resolves the "updated_at" problem (+ tests), taking a look on your second comment.

Reply:
> In HttpClient._handle_error_resolution, if error_resolution.response_action not in self._ACTIONS_TO_RETRY_ON, then evict the entry from the cache where self._ACTIONS_TO_RETRY_ON = {ResponseAction.RETRY, ResponseAction.RATE_LIMITED}

Ok, I got it. This is good as we would usually keep a light store. It makes most of my test useless, but I will remove it for now while I get a new one.

Reply:
I think your comment would indeed solve the issue. Thinking again about it, I think the second solution is more interesting because:
* It is less complex in the sense that we don't need to maintain a new data structure
* It is more efficient as we don't have to check the new data structure on each HTTP request

I think there is a drawback which is that it isn't so clear that we are at risk of a memory leak if we don't have an explicit data structure that ensures that (so this would be a DX issue) but if we can get a good test to ensure this, I think I would be sold.

On your side, do you have a preference?

Reply:
> It is less complex in the sense that we don't need to maintain a new data structure

This could be a problem for unexpected behavior/scenarios.

> but if we can get a good test to ensure this, I think I would be sold.

Let me try to achieve this through the second solution

Reply:
Ok, I updated to second propsal which is cleaner and simpler IMO, also added unit test that keeps an eye on count store size.